src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -52,10 +51,17 @@ internal int FindMetricAggregators(ReadOnlySpan<KeyValuePair<string, object>> ta
             int len = tags.Length;
             if (len == 0)
             {
-                if (this.metrics[0].StartTime == default)
+                if (!this.zeroTagMetricPointInitialized)
                 {
-                    var dt = DateTimeOffset.UtcNow;
-                    this.metrics[0] = new MetricPoint(this.aggType, dt, null, null);
+                    lock (this.lockZeroTags)
+                    {
+                        if (!this.zeroTagMetricPointInitialized)
+                        {
+                            var dt = DateTimeOffset.UtcNow;
```

Comment:
Do we want to put these logics in the lock or not? Need to find a good balance.
Think about:
1. if we put the logic inside the lock, it'll have the run-once guarantee, but the contention is becoming bigger so there is high chance for the lock to be escalated from a fast CPU spin to a real OS synchronization object.
2. if we put the logic before the lock, we might run the code more than once (e.g. getting UtcNow multiple times), but it won't cause other side effects besides burning some extra CPU cycles.
3. if we put the logic after the lock, we might be able to use some local variable to indicate if we need to run it or not, so we have the exact run-once guarantee, however it might introduce stale reads to other threads depending on the design between readers and writers.

Reply:
Good points @reyang. TBH, I'm not sure. I _feel_ like this is OK because we will only hit the contention on cold start with multiple 0-length tags racing?

Some other kind of crazy perf-theory to consider...

Let's say everything under `if (!this.zeroTagMetricPointInitialized)` is uncommon path. We should move it into its own method and add `MethodImplOptions.NoInlining` on it. The reason is we ideally want our hot-path code to fit into what the CPU can load into its instruction cache. Something like that I see this happening in dotnet/runtime a lot 😅

Reply:
Yeah, just want to point it out so whatever decision we made here will be a conscious decision if it is on a critical code path, rather than saying "I don't know, never thought about that before".

Reply:
Thanks for the ideas. Tracking it under https://github.com/open-telemetry/opentelemetry-dotnet/issues/2374, so that these comments are not lost.